Issue #8186: add NoWhitespaceBeforeCaseDefaultColonCheck

wilcoln opened the pull request:
Resolve #8186 

- Diff report : https://wilcoln.github.io/checkstyle-reports/8186/diff/index.html
- Contrib repo PR : https://github.com/checkstyle/contribution/pull/471

wilcoln commented:
@romani, please review

strkkk commented on src/main/resources/google_checks.xml:
```diff
@@ -260,6 +260,10 @@
         </module>
+        <module name="NoWhitespaceBeforeCaseDefaultColon">
+            <property name="tokens"
+             value="COLON"/>
```
I see no reason to add tokens property

wilcoln replied:
Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,18 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespacebeforecasedefaultcolon;
+
+public class InputNoWhitespaceBeforeCaseDefaultColon {
```
Config comment

wilcoln replied:
Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,18 @@
+                break;
+            case 2:
+                break;
+            default : // violation
```
should be more cases, like
1. OK code
2. nested case/default (should not cause problem, but just in case)
3. colon on next line
4. comment between case/default and colon

wilcoln replied:
Done.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheckTest.java:
```diff
@@ -0,0 +1,57 @@
+    public void testTokensNotNull() {
+        final NoWhitespaceBeforeCaseDefaultColonCheck check =
+                new NoWhitespaceBeforeCaseDefaultColonCheck();
+        assertNotNull(check.getAcceptableTokens(), "Acceptable tokens should not be null");
```
`assertNotNull` this is not ok, please check that tokens return array with 1 elemnt COLON

wilcoln replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,99 @@
+ * <pre>
+ * {
+ *  switch(1) {
+ *     case 1 : // expected violation
```
// expected violation -> just // violation

wilcoln replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,99 @@
+ *       break;
+ *     case 2:
+ *       break;
+ *     default : // expected violation
```
same here

wilcoln replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,99 @@
+ *     default : // expected violation
+ *       break;
+ *  }
+ * }
```
example of OK code is needed

wilcoln replied:
Done.

wilcoln replied:
Done.

strkkk replied:
I dont see it :)

wilcoln replied:
I thought the valid switch case (`case 2:`) was enough.
But I guess you want me to add another switch with all cases and default valid.
I'm doing that.

strkkk replied:
It is better to have `// OK` comment on ok code and yes, examples of both tokens. You cannot do this in one switch, so it is better to have 2 of them.

wilcoln replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,99 @@
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * {
```
please add enclosing class to make example compilable

wilcoln replied:
Done.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheckTest.java:
```diff
@@ -0,0 +1,64 @@
+        final NoWhitespaceBeforeCaseDefaultColonCheck check =
+                new NoWhitespaceBeforeCaseDefaultColonCheck();
+        final int[] acceptableTokens = check.getAcceptableTokens();
+        assertTrue(acceptableTokens.length == 1 && acceptableTokens[0] == TokenTypes.COLON,
```
assertArrayEquals

wilcoln replied:
Thanks. Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,52 @@
+
+        switch(4) {
+            case 4
+                    :
```
should be a violation comment here

wilcoln replied:
Done.

strkkk commented:
@wilcoln Please also dont forget about regression report, there may be some interesting cases for test inputs.

wilcoln commented:
@strkkk , diff report has been uploaded.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,108 @@
+ * }
+ * </pre>
+ *
+ * @since 8.32
```
8.32 has been released, it should be 8.33 now

wilcoln replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * Checks that ast is a colon of switch case or default.
+     * @param colonAst DetailAST to check.
+     * @return true if ast is a COLON and is child of LITERAL_CASE or LITERAL_DEFAULT.
```
` true if ast is a COLON` 
that is not correct

wilcoln replied:
Thanks. Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,52 @@
+
+         for (int var1 : new int[]{}) {}
+
+         for (int var2: new int[]{}) {}
```
if you want to show that this cases should not cause violation, ternary operator can be added here as well.

wilcoln replied:
Done.

pbludov commented:
@wilcoln This check should be in the Google style config:  https://checkstyle.org/google_style.html#a4.6.2

wilcoln commented:
> @wilcoln This check should be in the Google style config: https://checkstyle.org/google_style.html#a4.6.2

@pbludov, this PR comprises changes in `google_style.xml`
https://github.com/checkstyle/checkstyle/blob/0d8bf9ea9c69ac58ab30378adf82b4c230cdc8ea/src/xdocs/google_style.xml#L1003-L1009

https://github.com/checkstyle/checkstyle/blob/0d8bf9ea9c69ac58ab30378adf82b4c230cdc8ea/src/xdocs/google_style.xml#L1052-L1056

If there's another file that needs to be modified, please let me know.

pbludov commented:
> If there's another file that needs to be modified, please let me know.

Sorry, I missed this change.

wilcoln commented:
@romani, please review.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,108 @@
+        final String line = getLine(ast.getLineNo() - 1);
+        final int before = ast.getColumnNo() - 1;
+        if (isColonOfCaseOrDefault(ast) && (before == -1
+                || Character.isWhitespace(line.charAt(before)))) {
```
please work on AST only, we discourage usage of `getLine` method as much as possible.
you need to compare two AST nodes and make sure that they are on same line and have columnNo on +1 of previous AST column + previous AST text.

wilcoln replied:
Done. Thank you for all the guidance, applying requested changes was really simple.

romani replied:
> getExpressionCharSize

you do not need whole expression size, you need `getLastColumOf(ats)` so just find last AST node in subtree and do size on it + column of last AST node. Please do.

It will not be very expensive.
But I agree that original implementation is was quicker, BUT there are bunch of validations we can do by regexp match .... but they are not reliable, and life showed us this several times. first version of checkstyle was completely based on non AST parsing, all that validations were deprecated and we use more expensive parsing of java files to make AST (most time of checkstyle execution we spend in parsing of java). So I am so skeptical on `getLines()` , it works , but during time .... there become too much issues with it.

wilcoln replied:
Understood, thanks again for the guidance. Done.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,54 @@
+
+        switch(4) {
+            case 4
+                    : // violation
```
please make extra case for: on next line by in good column position relative to previos line:
```
case 4
      : // violation
```

wilcoln replied:
Done.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,54 @@
+        switch(3) {
+            case 3 /* space around*/ : // violation
+                break;
+            default/* no space around */:
```
to support this case when logic is based on AST only, you need to make Check to be comments aware, theres is special getter in base class that make Check receive all comments. 
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L85

wilcoln replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,138 @@
+        if (isColonOfCaseOrDefault(ast)) {
+            final DetailAST parent = ast.getParent();
+            boolean flag = !TokenUtil.areOnSameLine(parent, ast);
+            if (!flag) {
```
this is very hard to read ...
```
onSameLine = TokenUtil.areOnSameLine(parent, ast);
if(onSameLine) {
...
}
if (!onSameLine || whiteSpaceDetected) {
}
```

wilcoln replied:
Sorry for the discomfort, Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,138 @@
+                        offset = 1;
+                    }
+                    flag = ast.getColumnNo() != previousLastChild.getColumnNo()
+                            + previousLastChild.getText().length() + offset;
```
`whiteSpaceDetected = ast.getColumnNo() != getLastColumnXxxxxx(ast)`

please make login in `visitToken` human readable, with all abstractions/formulas to be moved to methods.
The same for block below.

wilcoln replied:
Done.

wilcoln commented:
@romani, please review.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,155 @@
+    public void visitToken(DetailAST ast) {
+        final DetailAST parent = ast.getParent();
+        final boolean astIsColonOfCase = parent.getType() == TokenTypes.LITERAL_CASE;
+        final boolean astIsColonOfDefault = parent.getType() == TokenTypes.LITERAL_DEFAULT;
```
please rename both variables to avoid prefix `ast`, it is not required, whole Check is about ASTs.

romani commented:
@wilcoln , please regenerate diff report on latest commit to make it absolutely clear that we are good to be merged.

wilcoln commented:
@romani I've done multiple attempts of regresssion but they fail for projects latest JDK and JDK 14.
So now I'm doing it for another JDK.

I have no idea of what the cause is, but the program freezes during the report generation phase for both projects.

wilcoln commented:
Regression reported this false positive
```
        switch (s) {
               case
                   4:
                    break;
           }
```
So I've added it to UTs and changed the code.

rnveach commented:
pitest is failing.

romani commented:
> I have no idea of what the cause is, but the program freezes during the report generation phase for both projects.

ok, please use any other jdk version.

romani commented:
@wilcoln ,please finish PR, let us know if there is some problems.

romani commented:
@wilcoln , do you have time to finish this PR ?

wilcoln commented:
> @wilcoln , do you have time to finish this PR ?

@romani
First of all, sorry for the very late reply.

I'll like to finish this PR, but right now my priorities are on my exams and apprenticeship.

If it's ok by you, I can try to finish it on late January. By then exams will be over.

If not, I'll understand if you have to close the PR or assign the related issue to someone else.
In any case, I'm coming back in late January to check on this, If issue still exists, I'll retake it.

Thanks.

romani commented:
@wilcoln , please finish PR.

romani commented:
@wilcoln , ping.

romani commented:
@wilcoln , can you finish PR ? 
or I can close it as abandoned ....

romani commented:
if no action during next week, I will close this PR as abandoned.

AkMo3 commented:
I would like to work this since it is almost completed. I would send a new PR for it.

romani commented:
@AkMo3 , thanks a lot for helping with abandoned PRs, please reuse this code to speedup completion of this task
